### ISSUE TITLE
None of the deserializers work when using the 'multi:softprob' objective with Sagemaker XGBoost

### ISSUE THREAD

**Issue by rohitgmathews** [aws/sagemaker-spark#54]
### System Information
- **Spark or PySpark**: PySpark
- **SDK Version**: Bundled version that comes with EMR 5.13.0
- **Spark Version**: Bundled version that comes with EMR 5.13.0
- **Algorithm**: XGBoost (Classifier)

### Describe the problem
When using the 'multi:softprob' objective with Sagemaker the end point returns an array of doubles with the probability for each class. However, the default deserializer for XGBoost (XGBoostCSVRowDeserializer) tries to interpret the result as a Double when it tries to construct the dataframe in PySpark. This obviously fails with a NumberFormatException. Here is a stack trace.

```
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/dataframe.py", line 350, in show
  File "/usr/lib/spark/python/lib/py4j-0.10.6-src.zip/py4j/java_gateway.py", line 1160, in __call__
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/utils.py", line 63, in deco
  File "/usr/lib/spark/python/lib/py4j-0.10.6-src.zip/py4j/protocol.py", line 320, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o614.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 21.0 failed 1 times, most recent failure: Lost task 0.0 in stage 21.0 (TID 419, localhost, executor driver): java.lang.NumberFormatException: For input string: "[0.9360387325286865"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at scala.collection.immutable.StringLike$class.toDouble(StringLike.scala:284)
	at scala.collection.immutable.StringOps.toDouble(StringOps.scala:29)
	at com.amazonaws.services.sagemaker.sparksdk.transformation.deserializers.XGBoostCSVRowDeserializer$$anonfun$deserializeResponse$1.apply(XGBoostCSVRowDeserializer.scala:41)
	at com.amazonaws.services.sagemaker.sparksdk.transformation.deserializers.XGBoostCSVRowDeserializer$$anonfun$deserializeResponse$1.apply(XGBoostCSVRowDeserializer.scala:41)
```

### Minimal repo / logs
Build an XGBoost estimator with the objective as 'multi:softprob'. Say,
```python
    estimator = XGBoostSageMakerEstimator(
        trainingInstanceType=args.instance_type,
        trainingInstanceCount=args.instance_count,
        endpointInstanceType="ml.m4.xlarge",
        endpointInitialInstanceCount=1,
        sagemakerRole=IAMRole(SAGEMAKER_ROLE)
    )
    estimator.setObjective('multi:softprob')
    estimator.setNumClasses(2)
    estimator.setNumRound(10)
```
This successfully builds the endpoint. However, when you try to predict using the endpoint that was built above it fails with the stack trace pasted above
```python
    xgboost_model = SageMakerModel(
        endpointInstanceType=None,
        endpointInitialInstanceCount=None,
        requestRowSerializer=LibSVMRequestRowSerializer(),
        responseRowDeserializer=XGBoostCSVRowDeserializer(),
        existingEndpointName=end_point_name_of_end_point_created_above,
        endpointCreationPolicy=EndpointCreationPolicy.DO_NOT_CREATE,
        prependResultRows=False
    )

    transformed_data = xgboost_model.transform(df)
```

- **Exact command to reproduce**:
Read above. The endpoint is correctly able to predict when I set the objective as 'multi:softmax'. However, the deserializer just fails when the objective is 'multi:softprob'. I tried using `LibSVMResponseRowDeserializer` for giggles since I trained it with libsvm data, and as expected it didn't work either.
None of the deserializers listed here seem to be relevant either
```python
from .deserializers import (
    ResponseRowDeserializer,
    XGBoostCSVRowDeserializer, ProtobufResponseRowDeserializer,
    KMeansProtobufResponseRowDeserializer, PCAProtobufResponseRowDeserializer,
    LDAProtobufResponseRowDeserializer,
    LinearLearnerRegressorProtobufResponseRowDeserializer,
    LinearLearnerBinaryClassifierProtobufResponseRowDeserializer,
    FactorizationMachinesBinaryClassifierDeserializer,
    FactorizationMachinesRegressorDeserializer,
    LibSVMResponseRowDeserializer)

__all__ = ['ResponseRowDeserializer',
           'XGBoostCSVRowDeserializer',
           'ProtobufResponseRowDeserializer',
           'PCAProtobufResponseRowDeserializer',
           'LDAProtobufResponseRowDeserializer',
           'KMeansProtobufResponseRowDeserializer',
           'LinearLearnerBinaryClassifierProtobufResponseRowDeserializer',
           'LinearLearnerRegressorProtobufResponseRowDeserializer',
           'FactorizationMachinesBinaryClassifierDeserializer',
           'FactorizationMachinesRegressorDeserializer',
           'LibSVMResponseRowDeserializer']
```

**Comment by laurenyu** [comment]
Thanks for using SageMaker! This seems to be related to #47, which we're currently working on solving.

**Comment by rohitgmathews** [comment]
Thanks Lauren. I see that he faced the same issue somewhere down in the comments. I think the same workaround might work for us too (Using binary:logistic) since we have only 2 classes at the moment. Eager to know when xgboost fixes the nature of the response :)

**Comment by nadiaya** [comment]
Closing this one in favor of #47
